cpp/src/ts/batched_kalman.cu

```diff
@@ -0,0 +1,820 @@
+
+//   // Only allocates when the pointer is uninitialized.
+//   void allocate_if_zero(double*& ptr, size_t size);
+
+//   bool orderEquals(int p, int q, int num_batches);
+
+//   // static void resize_if_zero(thrust::device_vector<double> v, size_t size) {
+//   //   if (v.size() == 0) {
+//   //     v.resize(size);
+//   //   }
+//   // }
+
+//   ~KalmanContext() noexcept(false);
+// };
+
+// //! A global context variable which saves allocations between invocations.
+// // KalmanContext* KALMAN_CTX = nullptr;
+
+// KalmanContext::KalmanContext(int p, int q, int num_batches, cumlHandle& handle)
+//   : m_p(p), m_q(q), m_num_batches(num_batches) {
+//   // d_ys = nullptr;
+//   d_Fs = nullptr;
+//   d_loglike = nullptr;
+//   d_sigma2 = nullptr;
+//   d_ar = nullptr;
+//   d_Tar = nullptr;
+//   d_ma = nullptr;
+//   d_Tma = nullptr;
+
+//   handle.setStream(0);
+//   allocator = handle.getDeviceAllocator();
+// }
+
+// void KalmanContext::allocate_if_zero(double*& ptr, size_t size) {
+//   if (ptr == nullptr) {
+//     ptr = (double*)allocator->allocate(sizeof(double) * size, 0);
+//   }
+// }
+
+// bool KalmanContext::orderEquals(int p, int q, int num_batches) {
+//   return (m_p == p) && (m_q == q) && (m_num_batches == num_batches);
+// }
+
+// KalmanContext::~KalmanContext() noexcept(false) {
+//   ////////////////////////////////////////////////////////////
+//   // free memory
+//   // if (d_ys != nullptr) allocator->deallocate(d_ys, 0, 0);
+//   if (d_Fs != nullptr) allocator->deallocate(d_Fs, 0, 0);
+//   if (d_sigma2 != nullptr) allocator->deallocate(d_sigma2, 0, 0);
+//   if (d_loglike != nullptr) allocator->deallocate(d_loglike, 0, 0);
+
+//   if (d_ar != nullptr) allocator->deallocate(d_ar, 0, 0);
+//   if (d_Tar != nullptr) allocator->deallocate(d_Tar, 0, 0);
+//   if (d_ma != nullptr) allocator->deallocate(d_ma, 0, 0);
+//   if (d_Tma != nullptr) allocator->deallocate(d_Tma, 0, 0);
+//   CUDA_CHECK(cudaPeekAtLastError());
+// }
+
+void process_mem_usage(double& vm_usage, double& resident_set) {
+  vm_usage = 0.0;
+  resident_set = 0.0;
+
+  // the two fields we want
+  unsigned long vsize;
+  long rss;
+  {
+    std::string ignore;
+    std::ifstream ifs("/proc/self/stat", std::ios_base::in);
+    ifs >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >>
+      ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >>
+      ignore >> ignore >> ignore >> ignore >> ignore >> ignore >> ignore >>
+      ignore >> vsize >> rss;
+  }
+
+  long page_size_kb = sysconf(_SC_PAGE_SIZE) /
+                      1024;  // in case x86-64 is configured to use 2MB pages
+  vm_usage = vsize / 1024.0;
+  resident_set = rss * page_size_kb;
+}
+
+//! Thread-local Matrix-Vector multiplication.
+template <int r>
+__device__ void Mv_l(double* A, double* v, double* out) {
+  for (int i = 0; i < r; i++) {
+    out[i] = 0.0;
+    for (int j = 0; j < r; j++) {
+      out[i] += A[i + j * r] * v[j];
+    }
```

Comment:
Not sure if compiler would be able to optimize this. So afaster approach would be to do:
```cpp
    double sum = 0.0;
    for (int j = 0; j < r; j++) {
      sum += A[i + j * r] * v[j];
    }
    out[i] = sum;
```